### ISSUE TITLE
[Issue]: ROCm 7.14 regression on torch.nn.functional.scaled_dot_product_attention

### ISSUE THREAD

**Issue by bitgamma** [ROCm/ROCm#6601]
### Problem Description

with PyTorch 2.11, calling torch.nn.functional.scaled_dot_product_attention as part of the decode step of Qwen Image VAE generates artifacts in the final image. This does not happen with the same version of PyTorch built with ROCm 7.13 (specifically I tried `rocm7.13.0a20260430`).

Beside the initial artifacts (which manifest as localized strips of broken pixels, often barely visible), if the program allocates new buffers (which were likely just released by the sdpa function) and writes to them, a subsequent call to sdpa produces very severe artifacts and most of the frame is black. Interestingly enough invoking `torch.cuda.empty_cache()` prevents this second issue, hinting at some issue with memory management.

My take on the problem is that the wrong implementation of the kernel is being selected, since Qwen Image VAE has a dim=384 which is probably above the limit of flash and memefficient implementations. To corroborate this, manually forcing the "math" backend fixes the issue.  Forcing flash shows the exact artifacts I was describing, while forcing memefficient shows much worse degradation. I did not, however look into ROCm code to confirm this, so take this as speculation based on the observed behavior.

### Operating System

EndeavourOS

### CPU

AMD RYZEN AI MAX+ 395 w/ Radeon 8060S

### GPU

gfx1151

### ROCm Version

7.14, 7.15

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

<details>
<summary>rocminfo --support output</summary>

```
ROCk module is loaded
=====================    
HSA System Attributes    
=====================    
Runtime Version:         1.21
Runtime Ext Version:     1.26
System Timestamp Freq.:  1000.000000MHz
Sig. Max Wait Duration:  18446744073709551615 (0xFFFFFFFFFFFFFFFF) (timestamp count)
Machine Model:           LARGE                              
System Endianness:       LITTLE                             
Mwaitx:                  ENABLED
XNACK enabled:           NO
DMAbuf Support:          YES
VMM Support:             YES
Fabric Support:          NO

==========               
HSA Agents               
==========               
*******                  
Agent 1                  
*******                  
  Name:                    AMD RYZEN AI MAX+ 395 w/ Radeon 8060S
  Uuid:                    CPU-XX                             
  Marketing Name:          AMD RYZEN AI MAX+ 395 w/ Radeon 8060S
  Vendor Name:             CPU                                
  Feature:                 None specified                     
  Profile:                 FULL_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        0(0x0)                             
  Queue Min Size:          0(0x0)                             
  Queue Max Size:          0(0x0)                             
  Queue Type:              MULTI                              
  Node:                    0                                  
  Device Type:             CPU                                
  Cache Info:              
    L1:                      49152(0xc000) KB                   
  Chip ID:                 0(0x0)                             
  ASIC Revision:           0(0x0)                             
  Cacheline Size:          64(0x40)                           
  Max Clock Freq. (MHz):   5187                               
  BDFID:                   0                                  
  Internal Node ID:        0                                  
  Compute Unit:            32                                 
  SIMDs per CU:            0                                  
  Shader Engines:          0                                  
  Shader Arrs. per Eng.:   0                                  
  WatchPts on Addr. Ranges:1                                  
  Memory Properties:       
  Features:                None
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: FINE GRAINED        
      Size:                    131156760(0x7d14b18) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 2                   
      Segment:                 GLOBAL; FLAGS: EXTENDED FINE GRAINED
      Size:                    131156760(0x7d14b18) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 3                   
      Segment:                 GLOBAL; FLAGS: KERNARG, FINE GRAINED
      Size:                    131156760(0x7d14b18) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 4                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    131156760(0x7d14b18) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
  ISA Info:                
*******                  
Agent 2                  
*******                  
  Name:                    gfx1151                            
  Uuid:                    GPU-XX                             
  Marketing Name:          AMD Radeon 8060S Graphics          
  Vendor Name:             AMD                                
  Feature:                 KERNEL_DISPATCH                    
  Profile:                 BASE_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        128(0x80)                          
  Queue Min Size:          64(0x40)                           
  Queue Max Size:          131072(0x20000)                    
  Queue Type:              MULTI                              
  Node:                    1                                  
  Device Type:             GPU                                
  Cache Info:              
    L1:                      32(0x20) KB                        
    L2:                      2048(0x800) KB                     
    L3:                      32768(0x8000) KB                   
  Chip ID:                 5510(0x1586)                       
  ASIC Revision:           0(0x0)                             
  Cacheline Size:          128(0x80)                          
  Max Clock Freq. (MHz):   2900                               
  BDFID:                   49664                              
  Internal Node ID:        1                                  
  Compute Unit:            40                                 
  SIMDs per CU:            2                                  
  Shader Engines:          2                                  
  Shader Arrs. per Eng.:   2                                  
  WatchPts on Addr. Ranges:4                                  
  Coherent Host Access:    FALSE                              
  Memory Properties:       APU
  Features:                KERNEL_DISPATCH 
  Fast F16 Operation:      TRUE                               
  Wavefront Size:          32(0x20)                           
  Workgroup Max Size:      1024(0x400)                        
  Workgroup Max Size per Dimension:
    x                        1024(0x400)                        
    y                        1024(0x400)                        
    z                        1024(0x400)                        
  Max Waves Per CU:        32(0x20)                           
  Max Work-item Per CU:    1024(0x400)                        
  Grid Max Size:           4294967295(0xffffffff)             
  Grid Max Size per Dimension:
    x                        4294967295(0xffffffff)             
    y                        65535(0xffff)                      
    z                        65535(0xffff)                      
  Max fbarriers/Workgrp:   32                                 
  Packet Processor uCode:: 35                                 
  SDMA engine uCode::      18                                 
  IOMMU Support::          None                               
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    122880000(0x7530000) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:2048KB                             
      Alloc Alignment:         4KB                                
      Accessible by all:       FALSE                              
    Pool 2                   
      Segment:                 GLOBAL; FLAGS: EXTENDED FINE GRAINED
      Size:                    122880000(0x7530000) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:2048KB                             
      Alloc Alignment:         4KB                                
      Accessible by all:       FALSE                              
    Pool 3                   
      Segment:                 GROUP                              
      Size:                    64(0x40) KB                        
      Allocatable:             FALSE                              
      Alloc Granule:           0KB                                
      Alloc Recommended Granule:0KB                                
      Alloc Alignment:         0KB                                
      Accessible by all:       FALSE                              
  ISA Info:                
    ISA 1                    
      Name:                    amdgcn-amd-amdhsa--gfx1151         
      Machine Models:          HSA_MACHINE_MODEL_LARGE            
      Profiles:                HSA_PROFILE_BASE                   
      Default Rounding Mode:   NEAR                               
      Default Rounding Mode:   NEAR                               
      Fast f16:                TRUE                               
      Workgroup Max Size:      1024(0x400)                        
      Workgroup Max Size per Dimension:
        x                        1024(0x400)                        
        y                        1024(0x400)                        
        z                        1024(0x400)                        
      Grid Max Size:           4294967295(0xffffffff)             
      Grid Max Size per Dimension:
        x                        2147483647(0x7fffffff)             
        y                        65535(0xffff)                      
        z                        65535(0xffff)                      
      FBarrier Max Size:       32                                 
    ISA 2                    
      Name:                    amdgcn-amd-amdhsa--gfx11-generic   
      Machine Models:          HSA_MACHINE_MODEL_LARGE            
      Profiles:                HSA_PROFILE_BASE                   
      Default Rounding Mode:   NEAR                               
      Default Rounding Mode:   NEAR                               
      Fast f16:                TRUE                               
      Workgroup Max Size:      1024(0x400)                        
      Workgroup Max Size per Dimension:
        x                        1024(0x400)                        
        y                        1024(0x400)                        
        z                        1024(0x400)                        
      Grid Max Size:           4294967295(0xffffffff)             
      Grid Max Size per Dimension:
        x                        2147483647(0x7fffffff)             
        y                        65535(0xffff)                      
        z                        65535(0xffff)                      
      FBarrier Max Size:       32                                 
*******                  
Agent 3                  
*******                  
  Name:                    aie2p                              
  Uuid:                    AIE-XX                             
  Marketing Name:          RyzenAI-npu5                       
  Vendor Name:             AMD                                
  Feature:                 AGENT_DISPATCH                     
  Profile:                 BASE_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        1(0x1)                             
  Queue Min Size:          64(0x40)                           
  Queue Max Size:          64(0x40)                           
  Queue Type:              SINGLE                             
  Node:                    0                                  
  Device Type:             DSP                                
  Cache Info:              
    L2:                      2048(0x800) KB                     
    L3:                      32768(0x8000) KB                   
  Chip ID:                 0(0x0)                             
  ASIC Revision:           0(0x0)                             
  Cacheline Size:          0(0x0)                             
  Max Clock Freq. (MHz):   0                                  
  BDFID:                   0                                  
  Internal Node ID:        0                                  
  Compute Unit:            0                                  
  SIMDs per CU:            0                                  
  Shader Engines:          0                                  
  Shader Arrs. per Eng.:   0                                  
  WatchPts on Addr. Ranges:0                                  
  Memory Properties:       
  Features:                AGENT_DISPATCH
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: KERNARG, COARSE GRAINED
      Size:                    131156760(0x7d14b18) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 2                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    65536(0x10000) KB                  
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:0KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 3                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    131156760(0x7d14b18) KB            
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Recommended Granule:4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
  ISA Info:                
*** Done *** 
```

</details>

**Comment by Eskuero** [comment]
EDITED COMMENT: WRONGLY ASSUMED ROOT CAUSE FROM https://github.com/Comfy-Org/ComfyUI/issues/15436

Whatever the issue was it is still reproducible up until the rocm nightlies of August 17th:

`pip install --index-url https://rocm.nightlies.amd.com/whl-multi-arch/ "rocm[devel]==10.1.0a20260817" "torch[device-gfx1201]==2.15.0a0+rocm10.1.0a20260817" "torchvision[device-gfx1201]==0.30.0a0+rocm10.1.0a20260817" "torchaudio==2.11.0a0+rocm10.1.0a20260817" "amd-torch-device-gfx1201" "triton==3.8.0+gitdf3f91dd.rocm10.1.0a20260817"`

But it seems to have been fixed after the 18th nightlies and any of these work reliably:

`pip install --index-url https://rocm.nightlies.amd.com/whl-multi-arch/ "rocm[devel]==10.1.0a20260818" "torch[device-gfx1201]==2.15.0a0+rocm10.1.0a20260818" "torchvision[device-gfx1201]==0.30.0a0+rocm10.1.0a20260818" "torchaudio==2.11.0a0+rocm10.1.0a20260818" "amd-torch-device-gfx1201" "triton==3.8.0+gitdf3f91dd.rocm10.1.0a20260818"`

`pip install --index-url https://rocm.nightlies.amd.com/whl-multi-arch/ "rocm[devel]==10.1.0a20260819" "torch[device-gfx1201]==2.15.0a0+rocm10.1.0a20260819" "torchvision[device-gfx1201]==0.30.0a0+rocm10.1.0a20260819" "torchaudio==2.11.0a0+rocm10.1.0a20260819" "amd-torch-device-gfx1201" "triton==3.8.0+gitdf3f91dd.rocm10.1.0a20260819"`

`pip install --index-url https://rocm.nightlies.amd.com/whl-multi-arch/ "rocm[devel]==10.1.0a20260820" "torch[device-gfx1201]==2.15.0a0+rocm10.1.0a20260820" "torchvision[device-gfx1201]==0.30.0a0+rocm10.1.0a20260820" "torchaudio==2.11.0a0+rocm10.1.0a20260820" "amd-torch-device-gfx1201" "triton==3.8.0+gitdf3f91dd.rocm10.1.0a20260820"`

Sorry about the off topic.

**Comment by nkulshre-amd** [comment]
@bitgamma Same underlying issue as https://github.com/pytorch/pytorch/issues/189849

You're right that hdim >256 (in this case =384) is not supported for Flash SDPA on arch `gfx1151`, hence Math fallback is required.
I've highlighted the workaround in source but it's missing on the latest AOTriton v0.12 (shipped with ROCm v7.14/7.15).

You might manually downgrade to [v0.11](github.com/ROCm/aotriton/releases/tag/0.11b) but expect to run into compatibility issues: https://github.com/ROCm/TheRock/issues/7315

Downgrading further to AOTriton v0.10 / ROCm v7.13 works since Math is the selected backend.

**Comment by nkulshre-amd** [comment]
@Eskuero different issue since this is not related to memory constraints

**Comment by nkulshre-amd** [comment]
Fixed with https://github.com/cyyever/pytorch/commit/fba88a5dd7c53d679cb82fd33cbe084c3d6b3f31 on tip and should be available in release/2.15
